crates/matrix-sdk-crypto/src/machine.rs

```diff
@@ -233,11 +234,18 @@ impl OlmMachine {
     ) -> StoreResult<Self> {
         let account = match store.load_account().await? {
             Some(a) => {
-                debug!(
-                    ed25519_key = a.identity_keys().ed25519.to_base64().as_str(),
-                    "Restored an Olm account"
-                );
-                a
+                if user_id != a.user_id() || device_id != a.device_id() {
+                    return Err(CryptoStoreError::MismatchedAccount {
+                        expected: (a.user_id().to_owned(), a.device_id().to_owned()),
+                        got: (user_id.to_owned(), device_id.to_owned()),
+                    });
+                } else {
+                    debug!(
+                        ed25519_key = a.identity_keys().ed25519.to_base64().as_str(),
```

Comment:
`.as_str()` doesn't seem necessary here (both `String` and `&str` implement `tracing::Value`).

Reply:
Yeah those are remnants of the time before vodozemac. Made the whole thing a bit nicer by introducing a new span https://github.com/matrix-org/matrix-rust-sdk/pull/1452/commits/f213445a305e9de859d785f6de5bc719895b3819.